src/pandas/types/numeric.h

```diff
@@ -44,14 +106,153 @@ class PANDAS_EXPORT IntegerArray : public NumericArray<TYPE> {
 
   bool owns_data() const override;
 
+  Status EnsureMutable();
+
+  // In the case where there are any nulls in this array,
+  // this function will be faster than calling GetNullCount() > 0.
+  bool HasNulls() const {
+    for (auto ii = 0; ii < this->length_; ++ii) {
+      if (IsNull(ii)) { return true; }
+    }
+    return false;
+  }
+
+  // Marks the value at the specified index as null. A pre-condition
+  // is that the valid_bits for this object have already been
+  // initialized.
+  void SetNull(int64_t i) {
+    // Pre-condition is that valid_bits_ is non-null because
+    // operations that would set them (e.g. - operator+=)
+    // would first create the valid_bits_ if needed
+    BitUtil::SetBit(
+        static_cast<MutableBuffer&>(*valid_bits_).mutable_data(), this->offset_ + i);
+  }
+
+  // Marks the value at the specified index as null. A pre-condition
+  // is that the valid_bits for this object have already been
+  // initialized.
+  void SetValid(int64_t i) {
+    // Pre-condition is that valid_bits_ is non-null because
+    // operations that would set them (e.g. - operator+=)
+    // would first create the valid_bits_ if needed
+    BitUtil::ClearBit(
+        static_cast<MutableBuffer&>(*valid_bits_).mutable_data(), this->offset_ + i);
+  }
+
+  // Returns whether or not the value at the specified index has
+  // been marked as null.
+  bool IsNull(int64_t i) const {
+    if (valid_bits_) {
+      // TODO: range check?
+      return BitUtil::GetBit(valid_bits_->data(), this->offset_ + i);
+    } else {
+      return false;
+    }
+  }
+
+  // Copy the list of nulls (aka valid bits) to a buffer
+  Status CopyNulls(int64_t length, std::shared_ptr<Buffer>* out) const {
+    // Pre-condition is that we already have valid_bits_ set
+    return valid_bits_->Copy(this->offset_, BitUtil::BytesForBits(length), out);
+  }
+
+  template <typename OTHER_TYPE>
+  IntegerArray<TYPE>& operator+=(const IntegerArray<OTHER_TYPE>& other) {
+    // TODO: check Status and throw exception
+    EnsureMutable();
+    auto length = std::min(this->length_, other.length());
```

Comment:
If you assume that all numeric types will have a null bitmap, then I think NumericArray in the current code would probably hold something equivalent to what you are suggesting with TypeData.

I'm not following your SFINAE code, but I think you are driving at a design that looks a bit more like this:

```cpp
template <typename KERNEL,
          typename LEFT_TYPE, 
          typename RIGHT_TYPE, 
          typename OUT_TYPE,  
          typename... KERNEL_ARGS>
void RunKernel(TypeData<LEFT_TYPE> const& left, 
               TypeData<RIGHT_TYPE> const& right,
               TypeData<OUT_TYPE>& out,                    
               KERNEL_ARGS&&... kernel_args)
{
    KERNEL kernel(std::forward<KERNEL_ARGS>(kernel_args)...);
    kernel(left, right, out);
}

class GenericKernel
{

public:

    template <typename LEFT_TYPE, typename RIGHT_TYPE, typename OUT_TYPE, typename FUNCTOR>
    void Run(const TypeData<LEFT_TYPE>& left, 
             FUNCTOR& functor,
             const TypeData<RIGHT_TYPE>& right,
             TypeData<OUT_TYPE>& out)
    {
        // Do a bitmap OR for nulls first and then do ffs
        // and branch 
        for (auto ii = 0; ii < left.length; ++ii)
        {
            functor(left.data[ii], right.data[ii], out.data[ii]);
        }
    }    

};

// This would get initialized once at startup and would 
// have information about the instruction set available
// on the machine (AVX vs. AVX2)
class SIMDConfiguration
{
};

template <typename LEFT_TYPE, typename RIGHT_TYPE>
class SIMDSinglePrecisionPossible;

template <typename LEFT_TYPE, typename RIGHT_TYPE>
class SIMDSinglePrecisionPossible : public std::false_type
{
};

template <>
class SIMDSinglePrecisionPossible<float, float> : public std::true_type
{
};

template <typename LEFT_TYPE, typename RIGHT_TYPE>
class SIMDDoublePrecisionPossible;

template <typename LEFT_TYPE, typename RIGHT_TYPE>
class SIMDDoublePrecisionPossible : public std::false_type
{
};

template <>
class SIMDDoublePrecisionPossible<double, double> : public std::true_type
{
};

template <typename LEFT_TYPE, typename RIGHT_TYPE>
class SIMDImpossible 
{
public:
    static constexpr auto value = !(SIMDDoublePrecisionPossible<LEFT_TYPE, RIGHT_TYPE>::value ||
                                    SIMDSinglePrecisionPossible<LEFT_TYPE, RIGHT_TYPE>::value);
};


class AddKernel : public GenericKernel
{

public:

    AddKernel(SIMDConfiguration& config) : config_(config)
    {
    }

    template <typename LEFT_TYPE,
              typename RIGHT_TYPE,
              typename OUT_TYPE>
    void operator()(const TypeData<LEFT_TYPE>& left, 
                    const TypeData<RIGHT_TYPE>& right,
                    TypeData<OUT_TYPE>& out)
    {
        RunOptimized<LEFT_TYPE, RIGHT_TYPE>(left, right, out);
    }

    template <typename LEFT_TYPE, typename RIGHT_TYPE, typename OUT_TYPE>
    typename std::enable_if<SIMDSinglePrecisionPossible<LEFT_TYPE, RIGHT_TYPE>::value>::type
    RunOptimized(const TypeData<LEFT_TYPE>& left, 
                 const TypeData<RIGHT_TYPE>& right,
                 TypeData<OUT_TYPE>& out)
    {
        // Use _mm_add_ps or _mm256_add_pd depending on 
        // config_, fallback to Run
    }

    template <typename LEFT_TYPE, typename RIGHT_TYPE, typename OUT_TYPE>
    typename std::enable_if<SIMDDoublePrecisionPossible<LEFT_TYPE, RIGHT_TYPE>::value>::type
    RunOptimized(const TypeData<LEFT_TYPE>& left, 
                 const TypeData<RIGHT_TYPE>& right,
                 TypeData<OUT_TYPE>& out)
    {
        // Use _mm_add_pd or _mm256_add_pd depending on 
        // config_, fallback to Run
    }

    template <typename LEFT_TYPE, typename RIGHT_TYPE, typename OUT_TYPE>
    typename std::enable_if<SIMDImpossible<LEFT_TYPE, RIGHT_TYPE>::value>::type
    RunOptimized(const TypeData<LEFT_TYPE>& left, 
                 const TypeData<RIGHT_TYPE>& right,
                 TypeData<OUT_TYPE>& out)
    {
        static auto add = [](auto const& left, auto const& right, auto& out)
            {
            out = left + right;
            };
        Run(left, add, right, out);
    }


private:

    SIMDConfiguration& config_;
                  
};

int main(int argc, char* argv[])
{
    // Just to see if it compiles
    TypeData<float> a;
    TypeData<int> b;
    TypeData<int> c;
    SIMDConfiguration config;
    RunKernel<AddKernel>(a, b, c, config);
}
```
I would expect FloatingArray and IntegerArray to control what operations are allowed using SFINAE, in the same fashion that I've currently implemented it. The main difference would be that those functions would just invoke RunKernel and would setup the correct return type based on the desired rules for promotion of types.